### PR TITLE
Add analyzer for RequiresAssemblyFilesAttribute

### PR DESCRIPTION
**Pull request by mateoatr**
This adds an analyzer for the proposed attribute, `RequiresAssemblyFilesAttribute`. Once (if) the attribute is landed in the runtime, the appropriate APIs should be annotated in the runtime libraries and suitable tests should be added in here.

### CODE REVIEW

File: src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs [hunk 1]
```diff
@@ -46,91 +48,44 @@ public override void Initialize (AnalysisContext context)
...
 					OperationAnalysisContext operationContext,
-					IMethodSymbol method,
-					Location location)
+					IMethodSymbol method)
 				{
-					AttributeData? requiresUnreferencedCode;
 					// If parent method contains RequiresUnreferencedCodeAttribute then we shouldn't report diagnostics for this method
 					if (operationContext.ContainingSymbol is IMethodSymbol &&
-						TryGetRequiresUnreferencedCodeAttribute (operationContext.ContainingSymbol.GetAttributes (), out requiresUnreferencedCode))
+						operationContext.ContainingSymbol.TryGetAttributeWithMessageOnCtor (FullyQualifiedRequiresUnreferencedCodeAttribute, out AttributeData? requiresUnreferencedCode))
```
Comment (tlakollo) [inline src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs:76]: I think we could use the same operationContext.ContainingSymbol.HasAttribute() function here

**Review by tlakollo — APPROVED** [review]
Looks good to me

File: src/ILLink.RoslynAnalyzer/ISymbolExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+namespace ILLink.RoslynAnalyzer
+{
+	static class ISymbolExtensions
+	{
+		/// <summary>
+		/// Returns true if symbol <see paramref="symbol"/> has an attribute with name <see paramref="attributeName"/>.
+		/// </summary>
+		internal static bool HasAttribute (this ISymbol symbol, string attributeName)
+		{
+			return symbol.GetAttributes ().Where (a => a?.AttributeClass?.Name == attributeName).Count () > 0;
```
Comment (agocke) [inline src/ILLink.RoslynAnalyzer/ISymbolExtensions.cs:17]: Hot path -- no LINQ

File: test/ILLink.RoslynAnalyzer.Tests/RequiresAssemblyFilesAnalyzerTests.cs [hunk 3]
```diff
@@ -3,22 +3,22 @@
...
-		static Task VerifySingleFileUnsupportedAnalyzer (string source, params DiagnosticResult[] expected)
+		static Task VerifyRequiresAssemblyFilesAnalyzer (string source, params DiagnosticResult[] expected)
 		{
 			// TODO: Remove this once we have the new attribute in the runtime.
 			source = @"namespace System.Diagnostics.CodeAnalysis
 {
 #nullable enable
 	[AttributeUsage (AttributeTargets.Method | AttributeTargets.Constructor, Inherited = false, AllowMultiple = false)]
-	public sealed class SingleFileUnsupportedAttribute : Attribute
+	public sealed class RequiresAssemblyFilesAttribute : Attribute
```
Comment (tlakollo) [inline test/ILLink.RoslynAnalyzer.Tests/RequiresAssemblyFilesAnalyzerTests.cs:19]: Are we changing the attribute to match the [new proposal](https://github.com/dotnet/runtime/issues/45705#issuecomment-755776365) with the parameterless constructor and extended attribute usage?

File: src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs [hunk 4]
```diff
@@ -15,8 +15,10 @@ namespace ILLink.RoslynAnalyzer
 	public class RequiresUnreferencedCodeAnalyzer : DiagnosticAnalyzer
 	{
 		public const string DiagnosticId = "IL2026";
+		const string RequiresUnreferencedCodeAttribute = nameof (RequiresUnreferencedCodeAttribute);
+		const string FullyQualifiedRequiresUnreferencedCodeAttribute = "System.Diagnostics.CodeAnalysis." + RequiresUnreferencedCodeAttribute;
 
-		private static readonly DiagnosticDescriptor s_rule = new DiagnosticDescriptor (
+		private static readonly DiagnosticDescriptor RequiresUnreferencedCodeRule = new DiagnosticDescriptor (
```
Comment (agocke) [inline src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs:21]: Is there a special coding convention for private fields in the linker repo? I believe the standard `dotnet` guidelines are `s_` prefix for static privates.

File: src/ILLink.RoslynAnalyzer/ISymbolExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+				if (attr.AttributeClass?.Name == attributeName)
+					return true;
+
+			return false;
+		}
+
+		internal static bool TryGetRequiresAssemblyFileAttribute (this ISymbol symbol, out AttributeData? attribute)
+		{
+			attribute = null;
+			if (symbol.GetAttributes ().FirstOrDefault (attr => attr.AttributeClass is { } attrClass &&
```
Comment (agocke) [inline src/ILLink.RoslynAnalyzer/ISymbolExtensions.cs:27]: The lambda here creates an allocation, and `FirstOrDefault` requires an additional delegate allocation.

File: src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs (hunk 4 above)
Comment (mateoatr) [reply src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs:21]: There's not AFAIK, but we can adhere to the standard dotnet guidelines.

File: src/ILLink.RoslynAnalyzer/ISymbolExtensions.cs (hunk 5 above)
Comment (mateoatr) [reply src/ILLink.RoslynAnalyzer/ISymbolExtensions.cs]: I removed all Linq usages in this file as well - PTAL.

**Pull request merged**